mythril/laser/ethereum/state/memory.py

```diff
@@ -0,0 +1,108 @@
+from typing import Union
+from z3 import (
+    Array,
+    BitVecSort,
+    Z3Exception,
+    simplify,
+    Store,
+    BitVecVal,
+    Extract,
+    Concat,
+    BitVecRef,
+    BoolRef,
+    If,
+)
+
+from mythril.laser.ethereum import util
+
+
+class Memory:
+    def __init__(self, name: str):
+        """
+        Constructor for Memory
+        :param name: value representing the name for the z3 array representing the memory
+        """
+        self._memory = Array(name, BitVecSort(256), BitVecSort(8))
+        self._stored_indices = set()
+
+    def get_word_at(self, index: int) -> Union[int, BitVecRef]:
+        """
+        Access a word from a specified memory index
+        :param index: integer representing the index to access
+        :return: 32 byte word at the specified index
+        """
+        try:
+            return util.concrete_int_from_bytes(
+                bytes([util.get_concrete_int(b) for b in self[index : index + 32]]), 0
+            )
+        except TypeError:
+            return simplify(
+                Concat(
+                    [
+                        b if isinstance(b, BitVecRef) else BitVecVal(b, 256)
+                        for b in self[index : index + 32]
+                    ]
+                )
+            )
+
+    def write_word_at(self, index: int, value: Union[int, BitVecRef, BoolRef]):
+        """
+        Writes a 32 byte word to memory at the specified index`
+        :param index: index to write to
+        :param value: the value to write to memory
+        """
+        try:
+            # Attempt to concretize value
+            _bytes = util.concrete_int_to_bytes(value)
```

Comment:
Unfortunately value is an element taken from stack, it can be either `ExprRef, int, bytes, or boolean`. If it is boolean this might fails(it can be also be boolean as we can see it in the is_zero opcode) .One way might be is to handle boolean in here, 
another way is to reject boolean types in the stack and also adding an assert there in `MachineStack::push()` to fix to `Union[int, BitVecRef, BoolRef]` .

Reply:
I'll just add support for bool here for now. I looked into requiring that only BitVecRefs are in the stack, and it's not trivial, so when/if it happens, it should probably be in a different PR.